tools/slang-embed/slang-embed.cpp

```diff
@@ -0,0 +1,197 @@
+// slang-embed.cpp
+
+// This file implements a simple utility for taking an input file
+// and embedding into a C++ source file as a `static const` array.
+
+// For now this utility uses plain C stdlib functionality rather
+// than depending on any of the utiltiies from the Slang project
+// libraries.
+//
+#ifdef _MSC_VER
+#pragma warning(disable: 4996)
+#endif
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+// The utility is implemented as a single `struct` type
+// that provides a context for the code. We do this as
+// an alternative to using global variables for passing
+// around options easily.
+//
+struct App
+{
+    char const* appName = "slang-embed";
+    char const* inputPath = nullptr;
+
+    void parseOptions(int argc, char** argv)
+    {
+        // Options are currently all specified by position,
+        // so the parsing logic is simplistic.
+
+        if( argc > 0 )
+        {
+            appName = *argv++;
+            argc--;
+        }
+
+        if( argc > 0 )
+        {
+            inputPath = *argv++;
+            argc--;
+        }
+
+        if( !inputPath || (argc != 0) )
+        {
+            fprintf(stderr, "usage: %s <inputPath>\n", appName);
+            exit(1);
+        }
+    }
+
+    void processInputFile()
+    {
+        // Note: Eventually we might support multiple input files in a
+        // single invocation of the tool, but for now we only have
+        // a single file to process.
+
+        // We open the input file in text mode because we are currently
+        // embedding textual source files. If/when this utility gets
+        // used for binary files another mode could be called for.
+        //
+        // (Alternatively, we might always use binary mode, but this
+        // could lead to a difference in the embedded bytes based on
+        // the line ending convention of the host platform)
+        //
+        FILE* inputFile = fopen(inputPath, "r");
+        if( !inputFile )
+        {
+            fprintf(stderr, "%s: error: failed to open '%s' for reading\n", appName, inputPath);
+            exit(1);
+        }
+
+        // We derive an output path simply by appending `.cpp` to the input path.
+        //
+        // TODO: If we start adding more complicated options, a `-o` option
+        // to specify a desired output path would be an obvious choice.
+        //
+        char* outputPath = (char*) malloc(strlen(inputPath) + strlen(".cpp") + 1);
```

Comment:
May I request that we don't leak even in these kind of run-once tools. The reason is that if we enable any memory testing tools, they could complain these leaks during the build process and cause the build to fail.

Reply:
Good point. I'll fix this up to properly clean up the scratch memory.